tests/_internal/compatibility/test_experimental.py

```diff
@@ -211,6 +213,92 @@ def foo(return_value: int = 1):
         foo(z=3)
 
 
+def test_experimental_field_warning():
+    @experimental_field(
+        "value",
+        group="test",
+        help="This is just a test, don't worry.",
+    )
+    class Foo(BaseModel):
+        value: int
+
+    with pytest.warns(
+        ExperimentalFeature,
+        match=(
+            "The field 'value' is experimental. This is just a test, "
+            "don't worry. The interface or behavior may change without warning, "
+            "we recommend pinning versions to prevent unexpected changes. "
+            "To disable warnings for this group of experiments, disable "
+            "PREFECT_EXPERIMENTAL_WARN_TEST."
+        ),
+    ):
+        assert Foo(value=2).value == 2
+
+
+def test_experimental_field_warning_no_warning_when_not_provided():
+    @experimental_field(
+        "value",
+        group="test",
+        help="This is just a test, don't worry.",
+    )
+    class Foo(BaseModel):
+        value: int = 1
+
+    assert Foo().value == 1
+
+
+def test_experimental_field_warning_when():
+    @experimental_field(
+        "value",
+        group="test",
+        help="This is just a test, don't worry.",
+        when=lambda x: x == 4,
+    )
+    class Foo(BaseModel):
+        value: int = 1
+
+    assert Foo(value=2).value == 2
+
+    with pytest.warns(
+        ExperimentalFeature,
+        match=(
+            "The field 'value' is experimental. This is just a test, "
+            "don't worry. The interface or behavior may change without warning, "
+            "we recommend pinning versions to prevent unexpected changes. "
+            "To disable warnings for this group of experiments, disable "
+            "PREFECT_EXPERIMENTAL_WARN_TEST."
+        ),
+    ):
+        assert Foo(value=4).value == 4
+
+
+def test_experimental_field_opt_in():
+    @experimental_field(
+        "value",
+        group="test",
+        help="This is just a test, don't worry.",
+        opt_in=True,
+    )
+    class Foo(BaseModel):
+        value: int = 1
+
+    with pytest.raises(ExperimentalFeatureDisabled):
+        assert Foo(value=1) == 1
+
+
+def test_experimental_field_retains_error_with_invalid_arguments():
+    @experimental_field(
+        "value",
+        group="test",
+        help="This is just a test, don't worry.",
+    )
+    class Foo(BaseModel):
+        value: int = 1
```

Comment:
Can this be implemented as this instead?
```suggestion
    class Foo(BaseModel):
        value: int = ExperimentalField(default=1, group="test", help="...")
```

It'd then add a validator to the field or something that raises the warning? It seems clearer perhaps?

Reply:
Or perhaps
```suggestion
    class Foo(BaseModel):
        value: int = ExperimentalField(Field(default=1, ...), group="test", help="...")
```

Reply:
I tried that initially, but I couldn't find a way to add validator by wrapping `Field`. The validators for fields seem to be place on the class that is used as the type hint. Do you have an example of extending `Field`?

Reply:
Ah interesting. Let's stick with this for now then.

Reply:
It looks like dynamically adding validators to an existing model isn't supported: https://github.com/pydantic/pydantic/issues/2076